src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -0,0 +1,127 @@
+        public Task<TResponseType> ExecuteAsync<TRequestType, TResponseType>(bool mutatesSolutionState, IRequestHandler<TRequestType, TResponseType> handler, TRequestType request,
+            ClientCapabilities clientCapabilities, string? clientName, CancellationToken cancellationToken) where TRequestType : class
+        {
+            // Create a task completion source that will represent the processing of this request to the caller
+            var completion = new TaskCompletionSource<TResponseType>();
+
+            var item = new QueueItem(mutatesSolutionState, clientCapabilities, clientName,
+                callback: async (context) =>
+                {
+                    // Check if cancellation was requested while this was waiting in the queue
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        completion.SetCanceled();
+                        return false;
+                    }
+
+                    try
+                    {
+                        var result = await handler.HandleRequestAsync(request, context, cancellationToken).ConfigureAwait(false);
+                        completion.SetResult(result);
+                        return true;
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        completion.SetCanceled();
+                    }
+                    catch (Exception exception)
+                    {
+                        // Pass the exception to the task completion source, so the caller of the ExecuteAsync method can observe but
+                        // don't let it escape from this callback, so it doens't affect the queue processing.
+                        completion.SetException(exception);
+                    }
+                    return false;
+                });
+            _queue.Enqueue(item);
+
+            return completion.Task;
+        }
+
+        private async Task ProcessQueueAsync()
+        {
+            while (true)
+            {
+                var work = await _queue.DequeueAsync().ConfigureAwait(false);
+
+                var solution = GetCurrentSolution();
+                solution = MergeChanges(solution, _lastMutatedSolution);
+
+                Solution? mutatedSolution = null;
+                var context = new RequestContext(solution, s => mutatedSolution = s, work.ClientCapabilities, work.ClientName);
+
+                if (work.MutatesSolutionState)
+                {
+                    // Mutating requests block other requests from starting to ensure an up to date snapshot is used.
+                    var ranToCompletion = await work.Callback(context).ConfigureAwait(false);
+
+                    // If the handling of the request failed, the exception will bubble back up to the caller, but we
+                    // still need to react to it here by throwing away solution updates
```

Comment:
need to think about this.  i thinki we should chat on what semantics we want ehre.  

1. if a mutation operation fails, i honestly don't know *waht* we can do.  because our state is literally out sync with whatever host we have.  contiuing will almost certainly lead to crashes.  i think we should just yellow-bar and shutdown there.

Reply:
The user will still see the yellow bar. I talked about this in a comment here: https://github.com/dotnet/roslyn/pull/46660#discussion_r467659997

What exactly to you mean by "shutdown"? I would want the queue to keep processing because closing and reopening the document should be all that is required to get back to a good state.

Reply:
> I would want the queue to keep processing because closing and reopening the document should be all that is required to get back to a good state.

1. do users know that?
2. in the meantime, we're out of sync, and i wouldn't expect to be able to handle *any* messages for that doc.  i.e. even a simple "what are the colors" request will likely lead to gibberish and/or crashes due to different bounds on the client/server.

Reply:
I don't disagree, but not not sure what we can do about it. The client could decide to respond to the error by closing/reopening the document (or re-initializing it) and then there wouldn't be any future problem. Maybe we could keep a "bad document list" and refuse to respond to requests for that document until we get a `didOpen`?

We should discuss more formally. This PR is to a feature branch so easy to follow up with changes later.

Reply:
I can't link the exact location in the spec, but if you search of inconsistent you should find it - https://microsoft.github.io/language-server-protocol/specification

> if servers end up in an inconsistent state they should log this to the client using the window/logMessage request. If they can’t recover from this the best they can do right now is to exit themselves. We are considering an extension to the protocol that allows servers to request a restart on the client side.

It sounds like we could exit, but if the LSP client doesn't attempt to restart us I'm not sure that's the best approach either.

Reply:
> We should discuss more formally.

Yup!

> Maybe we could keep a "bad document list" and refuse to respond to requests for that document until we get a didOpen?

I like this idea.  Including a way to inform users of this.  i.e. the yellow-bar, or console message can state explicitly that we've experienced a syncing problem with a doc and users can try to fix it by closing/opening.